tools/dhis2-audit-data-extractor/extract_audit.py

```diff
@@ -109,12 +67,133 @@ def extract_data():
+#         audit_data.append(json.loads(df.to_json(orient="records")))
 
-    print(json.dumps(data))
+#     return audit_data
 
+# @profile
+def extract_pgcopg2(format, output_mode, output_file, nr_rows, offset):
+    audit_data = list()
+    format = format.upper()
+    output_mode = output_mode.lower()
+    current_date = datetime.time(datetime.now())
+
+    conn = psycopg2.connect(
+        host=CONN_CONFIG['host'],
+        database=CONN_CONFIG['dbname'],
+        user=CONN_CONFIG['username'],
+        password=CONN_CONFIG['password'])
+
+    cur = conn.cursor()
+    cur.execute('SELECT * from audit ORDER BY createdat ASC LIMIT {} OFFSET {}'.format(nr_rows, offset))
+
+    if output_mode == "file":
+        filename = output_file if output_file else "dhis2_audit_extract-{0}.{1}".format(current_date, "csv" if format == "CSV" else "json")
+
+        with open(filename, 'a') as fd:
+            writer = csv.DictWriter(fd, fieldnames = ["id", "event", "type", "datetime", "createdby", "klass", "uid", "code", "attributes", "data"])
+            if format == "CSV" and not os.path.exists(filename):
+                writer.writeheader()
+
+            for row in iter_row(cur):
+                event = {
+                    "id": row[0],
+                    "event": row[1],
+                    "type": row[2],
+                    "datetime": row[3].strftime("%Y-%m-%d %H:%M:%S"),
+                    "createdby": row[4],
+                    "klass": row[5],
+                    "uid": row[6],
+                    "code": row[7],
+                    "attributes": row[8],
+                    "data": json.loads(gzip.decompress(row[9]).decode('utf-8')) if row[9] else None
+                }
+
+                if format == "CSV":
+                    writer.writerow(event)
+                elif format == "JSON":
+                    audit_data.append(event)
+
+            if format == "JSON":
+                json.dump(audit_data, fd)
+        print("Data saved in {0}".format(filename))
+    elif output_mode == "stdout":
+        for row in iter_row(cur):
+            if format == "CSV":
+                event = """{0},{1},{2},{3},{4},{5},{6},{7},'{8}','{9}'""".format(row[0],row[1],row[2],row[3].strftime("%Y-%m-%d %H:%M:%S"),row[4],row[5],row[6],row[7],row[8], json.loads(gzip.decompress(row[9]).decode('utf-8')) if row[9] else None)
+            elif format == "JSON":
+                event = {
+                    "id": row[0],
+                    "event": row[1],
+                    "type": row[2],
+                    "datetime": row[3].strftime("%Y-%m-%d %H:%M:%S"),
+                    "createdby": row[4],
+                    "klass": row[5],
+                    "uid": row[6],
+                    "code": row[7],
+                    "attributes": row[8],
+                    "data": json.loads(gzip.decompress(row[9]).decode('utf-8')) if row[9] else None
+                }
+            audit_data.append(event)
```

Comment:
I'd suggest extracting the event dictionary creation part into a separate function to allow the code to focus more on the 'what' than the 'how' of the output:

```suggestion
        def get_event(row):
            return {
                    "id": row[0],
                    "event": row[1],
                    "type": row[2],
                    "datetime": row[3].strftime("%Y-%m-%d %H:%M:%S"),
                    "createdby": row[4],
                    "klass": row[5],
                    "uid": row[6],
                    "code": row[7],
                    "attributes": row[8],
                    "data": json.loads(gzip.decompress(row[9]).decode('utf-8')) if row[9] else None
            }

        with open(filename, 'a') as fd:
            writer = csv.DictWriter(fd, fieldnames = ["id", "event", "type", "datetime", "createdby", "klass", "uid", "code", "attributes", "data"])
            if format == "CSV" and not os.path.exists(filename):
                writer.writeheader()

            for row in iter_row(cur):
                event = get_event(row)
                if format == "CSV":
                    writer.writerow(event)
                elif format == "JSON":
                    audit_data.append(event)

            if format == "JSON":
                json.dump(audit_data, fd)
        print("Data saved in {0}".format(filename))
    elif output_mode == "stdout":
        for row in iter_row(cur):
            event = get_event(row)
            if format == "CSV":
                line = """{0},{1},{2},{3},{4},{5},{6},{7},'{8}','{9}'""".format(
                    event.id, event.event, event.type, event.datetime, event.createdby, event.klass, event.uid, event.code, event.attributes, event.data
                )
            elif format == "JSON":
                line = event
            audit_data.append(line)
```

Reply:
This [23d05de](https://github.com/dhis2/dhis2-utils/pull/57/commits/23d05deff8e5cadc73c882db5eec6d4c611c2421) implements your suggestion.